### PR TITLE
generate a file for the elaborated boom config

### PR DESCRIPTION
**Pull request by ss2783**
This commit adds a file with the .core.config extension which saves
the same information as what was being printed out to STDOUT.

### CODE REVIEW

**Comment by ss2783** [comment]
There might be many things that are not scala idiomatic

Mostly works but I can't get rid of some extra empty braces in the execution unit prints. Here is an example:

<snip>
    -== Quad Issue ==-

     ExeUnit--
       - Mem
     ExeUnit--
       - ALU
       - Mul (pipelined)
       - Div
     ExeUnit--
       - ALU
       - IFPU (for read port access)
()         <------ can't get rid of this
     ExeUnit--
       - FPU (Latency: 4)
       - FDiv/FSqrt
       - FPIU (writes to Integer RF)
     ExeUnit--
       - IntToFP
<snio>

**Comment by ccelio** [comment]
Aren't you suppose to be working on your thesis??

**Comment by aswaterman** [comment]
Or busy leapfrogging from company to company

**Comment by ss2783** [comment]
🙌Thanks to RISC-V generated opportunities...juggling both 😀

File: src/main/scala/exu/core.scala [hunk 1]
```diff
@@ -222,44 +226,42 @@ class BoomCore(implicit p: Parameters, edge: freechips.rocketchip.tilelink.TLEdg
...
-   println("   Load/Store Unit Size  : " + NUM_LSU_ENTRIES + "/" + NUM_LSU_ENTRIES)
-   println("   Num Int Phys Registers: " + numIntPhysRegs)
-   println("   Num FP  Phys Registers: " + numFpPhysRegs)
-   println("   Max Branch Count      : " + MAX_BR_COUNT)
-   println("   BTB Size              : " +
+
+   override def toString: String =
+   ( exe_units_str + "\n"
+   + fp_pipeline_str + "\n"
+   + rob_str + "\n"
```
Comment (ccelio) [inline src/main/scala/exu/core.scala:235]: This is infuriating that rob.toString (and other similar modules) throws a `java.util.NoSuchElementException: head of empty list`. :(

I would like to think there is a way to do this without polluting the code with a bunch of forward declares.

File: src/main/scala/bpu/tage.scala [hunk 2]
```diff
@@ -222,9 +220,15 @@ class TageBrPredictor(
...
       if (i > 0) require(history_lengths(i) > history_lengths(i-1))
 
+      tables_str.append(table.toString + "\n")
       table
    }
 
+   override def toString: String =
+      ("  Building " + (size_in_bits/8/1024.0) + " kB TAGE Predictor (" +
+      + (size_in_bits/1024) + " Kbits) (max history length: " + history_lengths.max + " bits)\n"
+      + tables_str.toString)
```
Comment (ccelio) [inline src/main/scala/bpu/tage.scala:230]: You can instead use `tables.mkString("\n")` to do this instead. https://www.artima.com/pins1ed/next-steps-in-scala.html

**Review by ccelio — APPROVED** [review]
Thanks for doing this work. I'm going to clean it up a bit and merge it in now.

**Comment by ccelio** [comment]
Huh, it doesn't recognize I rebased this. Anyways, PR merged!

File: src/main/scala/exu/core.scala (hunk 1 above)
Comment (ccelio) [reply src/main/scala/exu/core.scala:235]: This looks like a Chisel bug. If I print out the core.toString everything works, but the ElaborationArtefacts() pukes on it, and it's on a chisel3.internal.naming.NamingStack.pop_return_context call. I'll escalate this elsewhere.

Comment (ccelio) [reply src/main/scala/exu/core.scala:235]: Oh, and I think this may be a `chiselName` macro error. Fun times.

**Pull request closed without merging**